### PR TITLE
[SPARK-47131][SQL][COLLATION] String function support: contains, startswith, endswith

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
Refactor built-in string functions to support collation for: contains, startsWith, endsWith.

### Why are the changes needed?
Add collation support for built-in string functions in Spark.

### Does this PR introduce _any_ user-facing change?
Yes, users should now be able to use COLLATE within arguments for built-in string functions: CONTAINS, STARTSWITH, ENDSWITH in Spark SQL queries.

### How was this patch tested?
Unit tests for:
- string expressions (StringExpressionsSuite)
- queries using "collate" (CollationSuite)

### Was this patch authored or co-authored using generative AI tooling?
Yes.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 1]
```diff
@@ -66,6 +66,13 @@ public static class Collation {
      */
     public final boolean isBinaryCollation;
 
+    /**
+     * Built-in string functions with collation support.
```
Comment (dbatomic) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:70]: Instead of adding new functions here I would propose following:
1) Add only low level basic functions here (e.g. matchAt).
2) Extend UTF8String with   `public boolean startsWith(final UTF8String prefix, *collationId*)`.  Have all the routing happen there.
3) Short circuit if we have `DEFAULT_COLLATION_ID` to previous implementation (e.g. (startsWith(final UTF8String prefix)). That way we will not risk regressing previous version.
4) Have operators call startsWith/endsWith... with proper collationId.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 2]
```diff
@@ -174,4 +174,291 @@ class CollationSuite extends QueryTest with SharedSparkSession {
...
+      checkAnswer(sql("SELECT startswith(collate('" + left + "', 'UCS_BASIC_LCASE'), collate('" +
+        right + "', 'UCS_BASIC_LCASE'))"), Row(expectedAnswer))
+      // UNICODE_CI
+      checkAnswer(sql("SELECT startswith('" + left + "', collate('" +
+        right + "', 'UNICODE_CI'))"), Row(expectedAnswer))
+      checkAnswer(sql("SELECT startswith(collate('" + left + "', 'UNICODE_CI'), collate('" +
+        right + "', 'UNICODE_CI'))"), Row(expectedAnswer))
+    }
+
+    // Serbian language collation tests
```
Comment (dbatomic) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:324]: Guess that this can be removed :)

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala]: Of course! Removed in new commit

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java (hunk 1 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:70]: Sounds good, on it!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 3]
```diff
@@ -586,9 +584,17 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
 }
 
 case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val collationID = right.dataType.asInstanceOf[StringType].collationId
```
[the remark this answers is not in the document]
Comment (mitkedb) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:588]: At the minimum, the codegen code for UTF8 binary should be identical before / after our collation changes

[the remark this answers is not in the document]
Comment (mitkedb) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:588]: Maybe we can have someone (e.g. @nikolamand-db ) build a java micro-benchmark in parallel to the UDF impl to assess the performance of the Java/JITed code, and we make decisions based on that?

[the remark this answers is not in the document]
Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:590]: Yeah, it does make sense to special case codegen here for DEFAULT_COLLATION_ID.

And also yes, adding micro-benchmark is on the list of tasks. We can figure out who is going to pick up that part of the work.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java (hunk 1 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java]: Update: as agreed in another comment thread, routing is now partially moved over to UTF8String, while binary collations are resolved right away in Scala in order to prevent performance regression

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Makes sense! Fixed

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 4]
```diff
@@ -105,6 +105,9 @@ public Collation(
   private static final Collation[] collationTable = new Collation[4];
   private static final HashMap<String, Integer> collationNameToIdMap = new HashMap<>();
 
+  public static final int DEFAULT_COLLATION_ID = 0;
```
Comment (dbatomic) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:108]: Let's avoid keeping COLLATION_IDs in both StringType and here. Maybe having an enum in CollationFactory is a good option?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 5]
```diff
@@ -341,6 +340,15 @@ public boolean contains(final UTF8String substring) {
     return false;
   }
 
+  public boolean contains(final UTF8String substring, int collationId) {
+    if (collationId == CollationFactory.LOWERCASE_COLLATION_ID) {
```
Comment (dbatomic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:344]: What if we pass DEFAULT_COLLATION_ID here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 6]
```diff
@@ -559,6 +557,19 @@ case class BinaryPredicate(override val prettyName: String, left: Expression, ri
   }
 }
 
+trait CollationCheck {
+  self: BinaryExpression =>
```
Comment (dbatomic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:561]: I think that you should implement this against `StringPredicate` instead of `BinaryExpression`. If it is just `BinaryExpression` you don't have any guarantee that `asInstanceOf[StringType]` will work.

Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:561]: Btw, good job adding this trait. We do need this!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 7]
```diff
@@ -622,10 +646,24 @@ object StartsWithExpressionBuilder extends StringBinaryPredicateExpressionBuilde
...
 }
 
-case class StartsWith(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.startsWith(r)
+case class StartsWith(left: Expression, right: Expression) extends StringPredicate
+  with CollationCheck {
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val collationId = checkCollationIds()
+    if (CollationFactory.fetchCollation(collationId).isBinaryCollation) {
+      l.startsWith(r)
```
Comment (dbatomic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:654]: Let's think about naming here. I don't like the idea that startsWith without collation implies binary check. Can we call these methods `startsWithBinary`/`containsBinary` etc.?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 8]
```diff
@@ -559,6 +557,19 @@ case class BinaryPredicate(override val prettyName: String, left: Expression, ri
...
 }
 
+trait CollationCheck {
+  self: BinaryExpression =>
+  def checkCollationIds(): Int = {
+    val leftCollationId: Int = left.dataType.asInstanceOf[StringType].collationId
+    val rightCollationId: Int = right.dataType.asInstanceOf[StringType].collationId
+    if (leftCollationId != rightCollationId) {
+      throw new IllegalArgumentException(
+        "Function requires the same collation type for left and right strings.")
```
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:567]: On it!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 7 above)
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:654]: You're right. "startsWith" was the original name for the binary method. Now, it would probably be best to change it to "startsWithBinary" (and probably "startsWithNonBinary" for the other one)

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 5 above)
Comment (dbatomic) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:344]: It's fine to have a short circuit and have `stringExpression` directly call `containsBinary`. I just think that  `contains` should work for all collations, including `UCS_BASIC`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Should be taken care of with new error classes (latest commit)

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 5 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:349]: Ahh, I see. Makes sense, I do agree that `contains(final UTF8String substring, int collationId)` should be able to handle all collations, even if it's not currently invoked for DEFAULT_COLLATION_ID. I added this in the latest commit!

However, that same default option that works for DEFAULT_COLLATION_ID (UCS_BASIC) also works for other binary collations supported by CollationFactory (currently: UNICODE).

I suppose we should go ahead and cover them all here using:
`if (CollationFactory.fetchCollation(collationId).isBinaryCollation)`
instead of just
` if (collationId == CollationFactory.DEFAULT_COLLATION_ID)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 7 above)
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Going with:
(old / current implementation - binary collations only) `containsBinary(substring)`
(new implementation - can be used for all collations) `contains(substring, collationId)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 9]
```diff
@@ -559,6 +559,25 @@ case class BinaryPredicate(override val prettyName: String, left: Expression, ri
   }
 }
 
+trait CollationCheck {
```
Comment (mkaravel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:562]: Rename to `CollationCompatibilityCheck` maybe to make it more clear what it does?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 10]
```diff
@@ -327,11 +329,10 @@ public UTF8String substringSQL(int pos, int length) {
   /**
    * Returns whether this contains `substring` or not.
    */
-  public boolean contains(final UTF8String substring) {
+  public boolean containsBinary(final UTF8String substring) {
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:332]: similar to `equals`/`compareTo`, let's not rename existing functions for backward compatibility, just add comments to say it's the binary version.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 11]
```diff
@@ -105,6 +105,9 @@ public Collation(
...
+  public static final int LOWERCASE_COLLATION_ID = 1;
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:109]: shall we use these constants in places like `collationTable[0]`?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 12]
```diff
@@ -341,6 +342,21 @@ public boolean contains(final UTF8String substring) {
...
+      return this.containsBinary(substring);
+    }
+    if (collationId == CollationFactory.LOWERCASE_COLLATION_ID) {
+      return this.toLowerCase().containsBinary(substring.toLowerCase());
+    }
+    // TODO: enable ICU collation support for "contains"
+    Map<String, String> params = new HashMap<>();
+    params.put("functionName", Thread.currentThread().getStackTrace()[1].getMethodName());
+    params.put("collationName", CollationFactory.fetchCollation(collationId).collationName);
+    throw new SparkException("COLLATION_NOT_SUPPORTED_FOR_FUNCTION",
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:356]: is this temporary?

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:356]: yes, will implement "contains" for non-binary collations in a future PR

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java (hunk 11 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:109]: for now, `DEFAULT_COLLATION_ID ` is mostly used outside of CollationFactory (e.g. StringType, ColumnType, PhysicalDataType), so there's not too many places where a "0" is hardcoded instead

while I suppose we could also do `collationTable[DEFAULT_COLLATION_ID] = ...`, it wouldn't really make too much sense as only `0` and `1` are special cases (while `2`, `3`, and so on are not), and I also suppose collationTable initialization itself will evolve soon with respect to other collation types so it hardly matters

that said, I could change it for now?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 12 above)
Comment (cloud-fan) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:356]: then we can throw an internal error for now, without an error class.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java (hunk 11 above)
Comment (cloud-fan) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:109]: got it, we can leave it for now.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 12 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:356]: I think the logic here was that there are gonna be some functions that are not (yet) supported for some particular collation types (some might get added at one point in the near future, while others may not), so we will eventually need a way to let users know. that's why we added the `COLLATION_NOT_SUPPORTED_FOR_FUNCTION` error class - while it will likely get removed from here, there are other places where it will appear instead

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 13]
```diff
@@ -331,7 +333,6 @@ public boolean contains(final UTF8String substring) {
     if (substring.numBytes == 0) {
       return true;
     }
-
```
Comment (dbatomic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:334]: Let's avoid unneeded changes.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 14]
```diff
@@ -341,6 +342,21 @@ public boolean contains(final UTF8String substring) {
...
+  public boolean contains(final UTF8String substring, int collationId) throws SparkException {
+    if (CollationFactory.fetchCollation(collationId).isBinaryCollation) {
+      return this.contains(substring);
+    }
+    if (collationId == CollationFactory.LOWERCASE_COLLATION_ID) {
+      return this.toLowerCase().contains(substring.toLowerCase());
+    }
+    // TODO: enable ICU collation support for "contains"
+    Map<String, String> params = new HashMap<>();
+    params.put("functionName", Thread.currentThread().getStackTrace()[1].getMethodName());
```
Comment (dbatomic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:354]: I would prefer simple `contains` instead of pulling function name from stack :)

Comment (cloud-fan) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:354]: We should use the SQL function name, which is `contains`. It doesn't matter what the internal java function name is to end users.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 15]
```diff
@@ -355,14 +371,43 @@ public boolean matchAt(final UTF8String s, int pos) {
...
   }
 
+  public boolean startsWith(final UTF8String prefix, int collationId) {
+    if (CollationFactory.fetchCollation(collationId).isBinaryCollation) {
+      return this.startsWith(prefix);
+    }
+    if (collationId == CollationFactory.LOWERCASE_COLLATION_ID) {
+      return this.toLowerCase().startsWith(prefix.toLowerCase());
+    }
+    // ICU collation support
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:392]: is this a TODO?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 16]
```diff
@@ -585,10 +604,23 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
   }
 }
 
-case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+case class Contains(left: Expression, right: Expression) extends StringPredicate
+  with CollationCompatibilityCheck {
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val collationId = checkCollationIds()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:610]: collation id should be checked only once during analysis, not per row. Please check `Expression.checkInputDataTypes`

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 13 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:334]: got it, thanks!

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 15 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:393]: no, this one is implemented

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 14 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: agreed!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 16 above)
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: fixed!

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 17]
```diff
@@ -341,6 +343,21 @@ public boolean contains(final UTF8String substring) {
...
   }
 
+  public boolean contains(final UTF8String substring, int collationId) throws SparkException {
+    if (CollationFactory.fetchCollation(collationId).isBinaryCollation) {
+      return this.contains(substring);
+    }
+    if (collationId == CollationFactory.LOWERCASE_COLLATION_ID) {
+      return this.toLowerCase().contains(substring.toLowerCase());
+    }
+    // TODO: enable ICU collation support for "contains"
```
Comment (dbatomic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:353]: You can also create a task in JIRA for this and add task id in comment.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 15 above)
Comment (dbatomic) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:393]: I think that it is confusing. Either delete it or make it useful to reader (e.g. "fallback to matchAt that will do collation aware comparison through ICU library").

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:393]: since passing collationId to `matchAt` already implies that it's a collation-aware implementation, which then uses `semanticCompare` with a corresponding ICU comparator, I guess the comment makes no sense anyways. removing

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 18]
```diff
@@ -586,9 +608,21 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
 }
 
 case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val collationId: Int = left.dataType.asInstanceOf[StringType].collationId
```
Comment (dbatomic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:612]: Can you push collationId as class property, outside of of `compare`?

That way you can share it between `doGenCode` and `compare`  + it should be more efficient to do cast once then per row.

**Review by dbatomic — APPROVED** [review]
LGTM (plus a couple of minor comments).

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 17 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:353]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 18 above)
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:612]: makes sense, avoid duplicate eval

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 19]
```diff
@@ -61,6 +61,5 @@ class StringType private(val collationId: Int) extends AtomicType with Serializa
  */
 @Stable
 case object StringType extends StringType(0) {
-  val DEFAULT_COLLATION_ID = 0
   def apply(collationId: Int): StringType = new StringType(collationId)
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:64]: @dbatomic public API should take collation name instead of collation id.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 20]
```diff
@@ -183,6 +183,266 @@ class CollationSuite extends DatasourceV2SQLBase {
...
+    val left: String = "abc" // collate with 'UNICODE_CI'
+    val leftCollationName: String = "UNICODE_CI";
+    var right: String = null // collate with 'UNICODE'
+    val rightCollationName: String = "UNICODE";
+    // contains
+    right = left.substring(1, 2);
+    checkError(
+      exception = intercept[SparkException] {
+        spark.sql(s"SELECT contains(collate('$left', '$leftCollationName')," +
+          s"collate('$right', '$rightCollationName'))").collect()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:196]:
```suggestion
          s"collate('$right', '$rightCollationName'))")
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 21]
```diff
@@ -183,6 +183,266 @@ class CollationSuite extends DatasourceV2SQLBase {
...
+      true, true, false, false, false, false, false, false, false, false, false,  //   c
+      true, true, true, false, false, false, false, false, false, false, false,   // abc
+      true, true, false, true, false, false, false, false, false, false, false,   //   cde
+      true, false, false, false, true, false, false, false, false, false, false,  // abde
+      true, true, true, true, false, true, false, false, false, false, false,     // abcde
+      true, false, false, false, false, false, true, false, false, false, false,  //   C
+      true, false, false, false, false, false, true, true, false, false, false,   // ABC
+      true, false, false, false, false, false, true, false, true, false, false,   //   CDE
+      true, false, false, false, false, false, false, false, false, true, false,  // ABDE
+      true, false, false, false, false, false, true, true, true, false, true)     // ABCDE
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:256]: This is hard to review...  Can we just provide a few typical cases to test?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 22]
```diff
@@ -497,10 +497,33 @@ case class Lower(child: Expression)
...
 
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val checkResult = super.checkInputDataTypes()
+    if (checkResult.isFailure) {
+      return checkResult
+    }
+    // Additional check needed for collation compatibility
+    val rightCollationId: Int = right.dataType.asInstanceOf[StringType].collationId
+    if (collationId != rightCollationId) {
+      throw new SparkException(
```
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: thanks, fixed!

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 20 above)
Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala]: I think I need .collect() in this case, otherwise it doesn't produce an exception here

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala (hunk 19 above)
Comment (dbatomic) [reply sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:64]: Ack. We will follow up on this.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 21 above)
Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:268]: while it may seem a bit unusual, I think this matrix approach covers a broad spectrum of test cases and generally works really well for this set of functions - covering various edge-cases and different collation types (this was especially useful when debugging and experimenting with new collations)

when I first wrote it as a standard linear set of tests, it was much harder to see how and why these functions behvae the way they do with different collations, while this nicely aligned matrix gives a pretty clear explanation all in one place

Comment (dbatomic) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:268]: Can we do case class checks, similarly to what we were doing in `CollationFactorySuite.scala`?

```scala
  case class CollationTestCase[R](collationName: String, s1: String, s2: String, expectedResult: R)

  test("collation aware equality and hash") {
    val checks = Seq(
      CollationTestCase("UCS_BASIC", "aaa", "aaa", true),
      CollationTestCase("UCS_BASIC", "aaa", "AAA", false),
      CollationTestCase("UCS_BASIC", "aaa", "bbb", false),
      CollationTestCase("UCS_BASIC_LCASE", "aaa", "aaa", true),
      CollationTestCase("UCS_BASIC_LCASE", "aaa", "AAA", true),
      CollationTestCase("UCS_BASIC_LCASE", "aaa", "AaA", true),
      CollationTestCase("UCS_BASIC_LCASE", "aaa", "AaA", true),
      CollationTestCase("UCS_BASIC_LCASE", "aaa", "aa", false),
      CollationTestCase("UCS_BASIC_LCASE", "aaa", "bbb", false),
      CollationTestCase("UNICODE", "aaa", "aaa", true),
      CollationTestCase("UNICODE", "aaa", "AAA", false),
      CollationTestCase("UNICODE", "aaa", "bbb", false),
      CollationTestCase("UNICODE_CI", "aaa", "aaa", true),
      CollationTestCase("UNICODE_CI", "aaa", "AAA", true),
      CollationTestCase("UNICODE_CI", "aaa", "bbb", false))
```

Also, having so many test cases may indeed be overkill :) I think that:
1) empty string
2/3) single char upper + lower
4/5) multichar upper + lower
6) Seq that is not contained anywhere.

Would suffice.

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:268]: looks great, thanks everyone - fixed!

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 20 above)
Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala]: added a check in alaysis as well, so this is finally removed

**Comment by cloud-fan** [comment]
@uros-db  unfortunately it has code conflicts now

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 23]
```diff
@@ -497,10 +497,32 @@ case class Lower(child: Expression)
 abstract class StringPredicate extends BinaryExpression
   with Predicate with ImplicitCastInputTypes with NullIntolerant {
 
+  final val collationId: Int = left.dataType.asInstanceOf[StringType].collationId
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:500]: I think it should be a lazy val, in case we fail with java cast issue before the `checkInputDataTypes` is called.

Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:500]: makes sense! fixed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 24]
```diff
@@ -586,9 +608,38 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
...
 case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val checkResult = super.checkInputDataTypes()
+    if (checkResult.isFailure) {
+      return checkResult
+    }
+    // Additional check needed for collation support
+    if (!CollationFactory.fetchCollation(collationId).isBinaryCollation
+      && collationId != CollationFactory.LOWERCASE_COLLATION_ID) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:618]: @dbatomic we should unify the naming in followup. binary stable, binary comparable, default collation, etc.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 25]
```diff
@@ -234,6 +438,7 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
         }.nonEmpty)
       }
     }
+
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:441]: let's avoid unnecessary changes.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
